Replace x/net websocket with gorilla websocket

fosterseth opened the pull request:
golang.org/x/net is missing features or is deprecated, so replacing with a more active library, github.com/gorilla/websocket

issue https://github.com/ghjm/sockceptor/issues/5

ghjm commented on pkg/backends/websockets.go:
```diff
@@ -29,13 +29,9 @@ type WebsocketDialer struct {
 	wd := WebsocketDialer{
 		address:     address,
-		origin:      fmt.Sprintf("http://%s", addrURL.Host),
+		origin:      fmt.Sprintf("http://%s", address),
```
I think the `address` parameter supplied to this function is expected to be a URL.  So you probably still have to parse out the host part to make an origin.  Maybe the scheme also needs to be http or https based on whether `address`'s URL scheme is `ws` or `wss`.

ghjm commented on pkg/backends/websockets.go:
```diff
@@ -47,21 +43,19 @@ func NewWebsocketDialer(address string, tlscfg *tls.Config, extraHeader string,
-				cfg.TlsConfig = b.tlscfg
+				dialer.TLSClientConfig  = b.tlscfg
 			}
+			header := make(http.Header, 0)
```
This declaration is going to be shadowed by the `:=` a few lines below, inside the `if {}` block.  This is a go-ism - every curly brace block implies a lexical scope, so the second `:=` declares a new variable that only exists within the `if {}`.  I think you get the behavior you want by just deleting the second declaration (line 60), which will make the `header.Add` apply to the outer variable.

fosterseth replied:
fixed

ghjm commented on pkg/backends/websockets.go:
```diff
@@ -47,21 +43,19 @@ func NewWebsocketDialer(address string, tlscfg *tls.Config, extraHeader string,
 			}
-			conn, err := websocket.DialConfig(cfg)
+			header.Add(http.CanonicalHeaderKey("origin"), b.origin)
+			u := url.URL{Scheme: "ws", Host: b.address, Path: "/"}
```
Callers need to be able to specify `ws://` or `wss://` - we can't assume one or the other here.  That's why `address` needs to be a full URL.

ghjm commented on pkg/backends/websockets.go:
```diff
@@ -136,28 +132,25 @@ func newWebsocketSession(conn *websocket.Conn) *WebsocketSession {
 func (ns *WebsocketSession) Send(data []byte) error {
 	buf := ns.framer.SendData(data)
-	n, err := ns.conn.Write(buf)
+	err := ns.conn.WriteMessage(websocket.BinaryMessage, buf)
```
Does `WriteMessage()` take a whole message, which will then be presented in its entirety on the other side via `ReadMessage()`?  If so then we don't actually need `framer` at all here.

fosterseth replied:
yes it takes whole message, so I removed framer.

ghjm commented:
As part of this PR, we should remove the TODOs at the top of `pkg/backends/websockets.go`

ghjm commented:
I was thinking that in the case where the user does `receptor --node id=foo --ws-peer address=wss://bar.com` it will fail because when it gets to `NewWebsocketDialer`, tlscfg will be nil, so it will be trying to run an unencrypted connection to a `wss://` endpoint.  It would be more user friendly to use default TLS settings in this case.

We have a client TLS configuration named `default` that is always created, and is intended for situations like this.  What I'm thinking is we check in `WebSocketDialerCfg.Run` if `cfg.TLS==""` and the scheme is `wss://`, then when we call `GetClientTLSConfig`, we pass in `default` as the config name instead of `cfg.TLS`.

ghjm commented on pkg/backends/websockets.go:
```diff
@@ -3,19 +3,17 @@ package backends
-//TODO: Switch to better supported websockets library
-//TODO: TLS Server
-//TODO: HTTP/HTTPS proxy support
+//TODO:
```
No reason to leave this here

fosterseth replied:
fixed

fosterseth committed:
Replace x/net websocket with gorilla websocket

golang.org/x/net is missing features or is deprecated, so
replacing with a more active library, github.com/gorilla/websocket

issue #5
